File: src/connectome/board.mjs
/**
 * board.mjs — Pure deterministic match-3 engine as dependency-free ESM.
 *
 * Same semantics as grid-engine.js (kept byte-compatible on purpose), but
 * importable from browser <script type="module"> AND Node trainers.
 * The RL loop trains against THIS file; the game validates with it too,
 * so a policy bug can never corrupt board state (invalid swaps revert).
 */
export const ROWS = 8, COLS = 8, NCOL = 6;

export function mulberry32(seed) {
  let a = (seed >>> 0) || 1;
  return function () {
    a |= 0; a = (a + 0x6D2B79F5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export const cloneBoard = (b) => b.map((r) => r.slice());

export function createBoard(seed) {
  const rng = mulberry32(seed), b = [];
  for (let r = 0; r < ROWS; r++) {
    b.push([]);
    for (let c = 0; c < COLS; c++) {
      let v = 0, g = 0;
      do { v = Math.floor(rng() * NCOL); g++; if (g > 60) break; }
      while ((c >= 2 && b[r][c - 1] === v && b[r][c - 2] === v) ||
             (r >= 2 && b[r - 1][c] === v && b[r - 2][c] === v));
      b[r].push(v);
    }
  }
  return b;
}

export function findMatches(board) {
  try {
    if (!Array.isArray(board) || !board.length || !Array.isArray(board[0])) return [];
    const rows = board.length, cols = board[0].length, mk = new Set();
    const K = (r, c) => r * 1000 + c;
    for (let r = 0; r < rows; r++) {
      if (!Array.isArray(board[r])) continue;
      let run = 1;
      for (let c = 1; c <= cols; c++) {
        const cur = c < cols ? board[r][c] : null, prev = board[r][c - 1];
        if (c < cols && cur === prev && typeof cur === 'number') run++;
        else { if (run >= 3 && typeof prev === 'number') for (let k = c - run; k < c; k++) mk.add(K(r, k)); run = 1; }
      }
    }
    for (let c = 0; c < cols; c++) {
      let run = 1;
      for (let r = 1; r <= rows; r++) {
        const cur = (r < rows && Array.isArray(board[r])) ? board[r][c] : null;
        const prev = Array.isArray(board[r - 1]) ? board[r - 1][c] : undefined;
        if (r < rows && cur === prev && typeof cur === 'number') run++;
        else { if (run >= 3 && typeof prev === 'number') for (let k = r - run; k < r; k++) mk.add(K(k, c)); run = 1; }
      }
    }
    return [...mk].map((k) => [Math.floor(k / 1000), k % 1000]);
  } catch { return []; }
}

function collapseAndRefill(board, cells, rng) {
  const rows = board.length, cols = board[0].length;
  const gone = new Set(cells.map(([r, c]) => r * 1000 + c));
  for (let c = 0; c < cols; c++) {
    const col = [];
    for (let r = rows - 1; r >= 0; r--) if (!gone.has(r * 1000 + c)) col.push(board[r][c]);
    while (col.length < rows) col.push(Math.floor(rng() * NCOL));
    for (let r = rows - 1, i = 0; r >= 0; r--, i++) board[r][c] = col[i];
  }
}

export function scoreForMatch(n) {
  if (n >= 5) return 200 + (n - 5) * 60;
  if (n === 4) return 120;
  return 60;
}

export function resolveCascades(board, rng) {
  const b = cloneBoard(board);
  let score = 0, casc = 0, removed = 0;
  const steps = [];
  for (let i = 0; i < 50; i++) {
    const m = findMatches(b);
    if (!m.length) break;
    casc++; removed += m.length;
    const pts = scoreForMatch(m.length) * casc;
    score += pts;
    steps.push({ cascade: casc, removed: m.length, points: pts });
    collapseAndRefill(b, m, rng);
  }
  return { board: b, score, cascades: casc, removed, steps };
}

/** Swap two cells by FLAT index 0..63 + direction. Reverts when no match. */
export function tryAction(board, cell, dir, rng) {
  const bad = (reason) => ({ ok: false, reason, board: cloneBoard(board), score: 0, removed: 0 });
  try {
    const r1 = Math.floor(cell / 8), c1 = cell % 8;
    const d = { up: [-1, 0], down: [1, 0], left: [0, -1], right: [0, 1] }[dir];
    if (!d) return bad('bad-dir');
    const r2 = r1 + d[0], c2 = c1 + d[1];
    if (r2 < 0 || r2 >= ROWS || c2 < 0 || c2 >= COLS) return bad('off-board');
    const b = cloneBoard(board);
    const t = b[r1][c1]; b[r1][c1] = b[r2][c2]; b[r2][c2] = t;
    if (!findMatches(b).length) return { ok: false, reason: 'no-match', board: cloneBoard(board), score: 0, removed: 0 };
    const res = resolveCascades(b, rng);
    return { ok: true, reason: 'match', board: res.board, score: res.score, removed: res.removed, steps: res.steps };
  } catch { return bad('exception'); }
}

export function findValidMove(board) {
  try {
    if (!Array.isArray(board) || !board.length) return null;
    for (let r = 0; r < ROWS; r++) {
      if (!Array.isArray(board[r])) continue;
      for (let c = 0; c < COLS; c++) for (const [dr, dc] of [[0, 1], [1, 0]]) {
        const r2 = r + dr, c2 = c + dc;
        if (r2 >= ROWS || c2 >= COLS || !Array.isArray(board[r2])) continue;
        const t = board[r][c]; board[r][c] = board[r2][c2]; board[r2][c2] = t;
        const hit = findMatches(board).length > 0;
        board[r2][c2] = board[r][c]; board[r][c] = t;
        if (hit) return { r1: r, c1: c, r2, c2 };
      }
    }
    return null;
  } catch { return null; }
}

export function reshuffle(board, seed) {
  try {
    const flat = board.flat();
    for (let a = 0; a < 200; a++) {
      const rng = mulberry32((seed + a * 7919) >>> 0), arr = flat.slice();
      for (let i = arr.length - 1; i > 0; i--) {
        const j = Math.floor(rng() * (i + 1));
        const t = arr[i]; arr[i] = arr[j]; arr[j] = t;
      }
      const b = [];
      for (let r = 0; r < ROWS; r++) b.push(arr.slice(r * COLS, (r + 1) * COLS));
      if (!findMatches(b).length && findValidMove(b)) return { board: b };
    }
    return { board: createBoard((seed + 999) >>> 0), fallback: true };
  } catch { return { board: createBoard(1234), fallback: true }; }
}

/** Valid direction mask for a flat cell (edge cells lose off-board dirs). */
export function validDirs(cell) {
  const r = Math.floor(cell / 8), c = cell % 8;
  return {
    up: r > 0, down: r < 7, left: c > 0, right: c < 7,
  };
}
export const DIRS = ['up', 'down', 'left', 'right'];

File: flycrush_py/board.py
"""Board engine: pure deterministic match-3 (mirrors board.mjs semantics).

mulberry32 is bit-compatible with the JS version so seeded boards match.
"""
from __future__ import annotations

ROWS, COLS, NCOL = 8, 8, 6
DIRS = ("up", "down", "left", "right")
_DV = {"up": (-1, 0), "down": (1, 0), "left": (0, -1), "right": (0, 1)}
_MASK = 0xFFFFFFFF


def _u32(x: int) -> int:
    return x & _MASK


class Rng:
    """Bit-compatible mulberry32."""

    def __init__(self, seed: int):
        self.a = _u32(seed) or 1

    def random(self) -> float:
        self.a = _u32(self.a + 0x6D2B79F5)
        t = _u32((self.a ^ (self.a >> 15)) * ((self.a | 1) & _MASK))
        t = _u32(t + _u32((t ^ (t >> 7)) * ((t | 61) & _MASK))) ^ t
        return (_u32(t ^ (t >> 14))) / 4294967296.0

    def randint(self, n: int) -> int:
        return int(self.random() * n)


def clone_board(b):
    try:
        return [row[:] for row in b]
    except Exception:
        return []


def create_board(seed: int):
    rng = Rng(seed)
    b = []
    for r in range(ROWS):
        b.append([])
        for c in range(COLS):
            v, g = 0, 0
            while True:
                v = rng.randint(NCOL)
                g += 1
                if g > 60:
                    break
                horiz = c >= 2 and b[r][c - 1] == v and b[r][c - 2] == v
                vert = r >= 2 and b[r - 1][c] == v and b[r - 2][c] == v
                if not (horiz or vert):
                    break
            b[r].append(v)
    return b


def find_matches(board) -> list:
    try:
        if not board or not board[0]:
            return []
        rows, cols = len(board), len(board[0])
        marked = set()
        for r in range(rows):
            run = 1
            for c in range(1, cols + 1):
                cur = board[r][c] if c < cols else None
                prev = board[r][c - 1]
                if c < cols and cur == prev and isinstance(cur, int):
                    run += 1
                else:
                    if run >= 3 and isinstance(prev, int):
                        for k in range(c - run, c):
                            marked.add((r, k))
                    run = 1
        for c in range(cols):
            run = 1
            for r in range(1, rows + 1):
                cur = board[r][c] if r < rows else None
                prev = board[r - 1][c]
                if r < rows and cur == prev and isinstance(cur, int):
                    run += 1
                else:
                    if run >= 3 and isinstance(prev, int):
                        for k in range(r - run, r):
                            marked.add((k, c))
                    run = 1
        return list(marked)
    except Exception:
        return []


def _collapse(board, cells, rng: Rng):
    rows, cols = len(board), len(board[0])
    gone = set(cells)
    for c in range(cols):
        col = [board[r][c] for r in range(rows - 1, -1, -1) if (r, c) not in gone]
        while len(col) < rows:
            col.append(rng.randint(NCOL))
        for r, i in zip(range(rows - 1, -1, -1), range(len(col))):
            board[r][c] = col[i]


def score_for_match(n: int) -> int:
    if n >= 5:
        return 200 + (n - 5) * 60
    if n == 4:
        return 120
    return 60


def resolve_cascades(board, rng: Rng, max_steps: int = 50):
    b = clone_board(board)
    score = casc = removed = 0
    steps = []
    for _ in range(max_steps):
        m = find_matches(b)
        if not m:
            break
        casc += 1
        removed += len(m)
        pts = score_for_match(len(m)) * casc
        score += pts
        steps.append({"cascade": casc, "removed": len(m), "points": pts})
        _collapse(b, m, rng)
    return {"board": b, "score": score, "cascades": casc, "removed": removed, "steps": steps}


def try_action(board, cell: int, direction: str, rng: Rng):
    def bad(reason):
        return {"ok": False, "reason": reason, "board": clone_board(board), "score": 0, "removed": 0}

    try:
        r1, c1 = divmod(cell, 8)
        d = _DV.get(direction)
        if d is None:
            return bad("bad-dir")
        r2, c2 = r1 + d[0], c1 + d[1]
        if not (0 <= r2 < ROWS and 0 <= c2 < COLS):
            return bad("off-board")
        b = clone_board(board)
        b[r1][c1], b[r2][c2] = b[r2][c2], b[r1][c1]
        if not find_matches(b):
            return {"ok": False, "reason": "no-match", "board": clone_board(board), "score": 0, "removed": 0}
        res = resolve_cascades(b, rng)
        return {"ok": True, "reason": "match", **res}
    except Exception:
        return bad("exception")


def find_valid_move(board):
    try:
        if not board:
            return None
        for r in range(ROWS):
            for c in range(COLS):
                for dr, dc in ((0, 1), (1, 0)):
                    r2, c2 = r + dr, c + dc
                    if r2 >= ROWS or c2 >= COLS:
                        continue
                    board[r][c], board[r2][c2] = board[r2][c2], board[r][c]
                    hit = bool(find_matches(board))
                    board[r2][c2], board[r][c] = board[r][c], board[r2][c2]
                    if hit:
                        return {"r1": r, "c1": c, "r2": r2, "c2": c2}
        return None
    except Exception:
        return None


def reshuffle(board, seed: int):
    try:
        flat = [v for row in board for v in row]
        for a in range(200):
            rng = Rng((seed + a * 7919) & _MASK)
            arr = flat[:]
            for i in range(len(arr) - 1, 0, -1):
                j = rng.randint(i + 1)
                arr[i], arr[j] = arr[j], arr[i]
            b = [arr[r * COLS:(r + 1) * COLS] for r in range(ROWS)]
            if not find_matches(b) and find_valid_move(b):
                return {"board": b}
        return {"board": create_board((seed + 999) & _MASK), "fallback": True}
    except Exception:
        return {"board": create_board(1234), "fallback": True}


def valid_dirs(cell: int) -> dict:
    r, c = divmod(cell, 8)
    return {"up": r > 0, "down": r < 7, "left": c > 0, "right": c < 7}
